lib/credentials/CredentialsManager.js

```diff
@@ -0,0 +1,170 @@
+                extension: this._extension,
+            });
+            return null;
+        }
+
+        const { id, accountId, stsclient, authConfig } = params;
+        this._accountCredsCache[id] = new VaultServiceCredentials(
+            stsclient,
+            this._extension,
+            getRoleArn(accountId, authConfig.vault.roleName),
+            this._logger.newRequestLogger());
+        return this._accountCredsCache[id];
+    }
+
+    _addAccountCredentials(params) {
+        const err = CredentialsManager.validateParams(params, accountParamsJoi);
+        if (err) {
+            this._logger.error('missing required params for account type credentials', {
+                method: 'CredentialsManager::_addVaultServiceCredentials',
+                error: err,
+                extension: this._extension,
+            });
+            return null;
+        }
+
+        const { id, authConfig } = params;
+        let credentials;
+
+        try {
+            credentials = new AccountCredentials(authConfig, this._logger);
+        } catch (err) {
+            this._logger.error(`unable get credentials for ${authConfig.account}`, {
+                method: 'CredentialsManager::_addAccountCredentials',
+                error: err,
+                extension: this._extension,
+            });
+            return null;
+        }
+
+        this._accountCredsCache[id] = credentials;
+        return this._accountCredsCache[id];
+    }
+
+    _addRoleCredentials(params) {
+        const err = CredentialsManager.validateParams(params, accountParamsJoi);
+        if (err) {
+            this._logger.error('missing required params for role type credentials', {
+                method: 'CredentialsManager::_addVaultServiceCredentials',
+                error: err,
+                extension: this._extension,
+            });
+            return null;
+        }
+
+        const { id, vaultclient, roleName } = params;
+        this._accountCredsCache[id] = new RoleCredentials(
+            vaultclient,
+            this._extension,
+            roleName,
+            this._logger.newRequestLogger());
+        return this._accountCredsCache[id];
+    }
+
+    getCredentials(params) {
+        const { authConfig, id } = params;
+
+        if (!id || !authConfig) {
+            this._logger.error('missing required params', {
+                method: 'CredentialsManager::getCredentials',
+                extension: this._extension,
+            });
+            return null;
+        }
+
+        if (this._accountCredsCache[id]) {
+            return this._accountCredsCache[id];
+        }
+
+        switch (authConfig.type) {
+        case authTypeRole:
+            return this._addRoleCredentials(params);
+        case authTypeVault:
+            return this._addVaultServiceCredentials(params);
+        case authTypeAccount: // fallthrough
+        case authTypeService:
+            return this._addAccountCredentials(params);
```

Comment:
Is there an immediate need to support multiple credentials types and all the overhead they incur? It seems like there is code that's not going to be used in the immediate future.

Reply:
Perhaps it's just the naming that confuses me, but why would the sts client need account credentials? Any access key would do provided it's linked to an entity with enough privileges.

By role, do you mean assumed role (the temp credentials issued by AssumeRole that include a session token), or the actual role name and access/secret key that will be used to assume a role?

Reply:
> Perhaps it's just the naming that confuses me, but why would the sts client need account credentials? Any access key would do provided it's linked to an entity with enough privileges. 

> By role, do you mean assumed role (the temp credentials issued by AssumeRole that include a session token), or the actual role name and access/secret key that will be used to assume a role?

I can see this being confusing now.
* `account` / `service`: will the credentials read from the `authdata.json` file. the sts client will be using this read the lifecycle credentials stored in the `authdata.json`

* ​`role`: at the moment, is the `assumeRoleBackbeat` credential type using the vaultclient

* `vault`:  will be the credential type using the new authentication with assume role we are currently implementing.

Reply:
May I suggest having the credential manager only handling type `vault` for now, and type `vault` renamed to `assumedRole` for clarity, given that:
- `role` is not used by lifecycle and not in scope of our current feature
- `account` / `service` do not need caching, refreshing or expiring and are readily available in configuration.

Reply:
* renamed `vault` to `assumeRole`
* renamed `VaultServiceCredentials` to `AssumeRoleCredentials` to better describe the class